subsys/lorawan/services/frag_transport.c

```diff
@@ -0,0 +1,310 @@
+
+static void frag_transport_package_callback(uint8_t port, bool data_pending, int16_t rssi,
+					    int8_t snr, uint8_t len, const uint8_t *rx_buf)
+{
+	uint8_t tx_buf[FRAG_TRANSPORT_MAX_CMDS_PER_PACKAGE * FRAG_TRANSPORT_MAX_ANS_LEN];
+	uint8_t tx_pos = 0;
+	uint8_t rx_pos = 0;
+	int ans_delay = 0;
+
+	__ASSERT(port == LORAWAN_PORT_FRAG_TRANSPORT, "Wrong port %d", port);
+
+	while (rx_pos < len) {
+		uint8_t command_id = rx_buf[rx_pos++];
+
+		if (sizeof(tx_buf) - tx_pos < FRAG_TRANSPORT_MAX_ANS_LEN) {
+			LOG_ERR("insufficient tx_buf size, some requests discarded");
+			break;
+		}
+
+		switch (command_id) {
+		case FRAG_TRANSPORT_CMD_PKG_VERSION:
+			tx_buf[tx_pos++] = FRAG_TRANSPORT_CMD_PKG_VERSION;
+			tx_buf[tx_pos++] = LORAWAN_PACKAGE_ID_FRAG_TRANSPORT_BLOCK;
+			tx_buf[tx_pos++] = FRAG_TRANSPORT_VERSION;
+			break;
+		case FRAG_TRANSPORT_CMD_FRAG_STATUS: {
+			uint8_t frag_status = rx_buf[rx_pos++] & 0x07;
+			uint8_t participants = frag_status & 0x01;
+			uint8_t index = frag_status >> 1;
+
+			LOG_DBG("FragSessionStatusReq index %d, participants: %u", index,
+				participants);
+
+			uint8_t missing_frag = CLAMP(ctx.nb_frag - ctx.nb_frag_received, 0, 255);
+
+			FragDecoderStatus_t decoder_status = FragDecoderGetStatus();
+			uint8_t memory_error = decoder_status.MatrixError;
+
+			if (participants == 1 || missing_frag > 0) {
+				tx_buf[tx_pos++] = FRAG_TRANSPORT_CMD_FRAG_STATUS;
+				tx_buf[tx_pos++] = ctx.nb_frag_received & 0xFF;
+				tx_buf[tx_pos++] =
+					(index << 6) | ((ctx.nb_frag_received >> 8) & 0x3F);
+				tx_buf[tx_pos++] = missing_frag;
+				tx_buf[tx_pos++] = memory_error & 0x01;
+
+				ans_delay = sys_rand32_get() % (1U << (ctx.block_ack_delay + 4));
+
+				LOG_DBG("FragSessionStatusAns index %d, NbFragReceived: %u, "
+					"MissingFrag: %u, MemoryError: %u, delay: %d",
+					index, ctx.nb_frag_received, missing_frag, memory_error,
+					ans_delay);
+			}
+			break;
+		}
+		case FRAG_TRANSPORT_CMD_FRAG_SESSION_SETUP: {
+			uint8_t frag_session = rx_buf[rx_pos++] & 0x3F;
+			uint8_t index = frag_session >> 4;
+			uint8_t status = index << 6;
+
+			if (!ctx.is_active || ctx.frag_index == index) {
+				ctx.frag_session = frag_session;
+				ctx.nb_frag_received = 0;
+
+				ctx.nb_frag = sys_get_le16(rx_buf + rx_pos);
+				rx_pos += sizeof(uint16_t);
+
+				ctx.frag_size = rx_buf[rx_pos++];
+				ctx.control = rx_buf[rx_pos++];
+				ctx.padding = rx_buf[rx_pos++];
+
+				ctx.descriptor = sys_get_le32(rx_buf + rx_pos);
+				rx_pos += sizeof(uint32_t);
+
+				LOG_INF("FragSessionSetupReq index %d, nb_frag: %u, frag_size: %u, "
+					"padding: %u, control: 0x%x, descriptor: 0x%.8x",
+					index, ctx.nb_frag, ctx.frag_size, ctx.padding, ctx.control,
+					ctx.descriptor);
+			} else {
+				/* FragIndex unsupported */
+				status |= 1U << 2;
+
+				LOG_WRN("FragSessionSetupReq failed. Session %u still active",
+					ctx.frag_index);
+			}
+
+			if (ctx.frag_algo > 0) {
+				/* FragAlgo unsupported */
+				status |= 1U << 0;
```

Comment:
Could these offsets be defined in an enum and set like so?
```
status |= BIT(ALGO_UNSUPPORTED)
```
The offsets are used in quite a few places

Reply:
I've just started defining these bits and sections as macros when I realized that bits and masks with the same meaning are not at the same positions depending on the message type. So this would lead to lots of lengthy macro definitions used exactly once. I find the code easier to read and easier to compare to the protocol specification if the numbers are directly visible in the code where the messages are assembled/disassembled.

I did follow your suggestion to use `BIT(0)` instead of `1U << 0` though, as this improves readability.